wagtail_transfer/richtext.py

```diff
@@ -0,0 +1,85 @@
+FIND_ID = re.compile(r'id="([^"]*)"')
+
+
+class RichTextReferenceHandler:
+    """
+    Handles updating ids and retrieving object references for <tag type_attribute="foo" id="my_id" /> tags representing references within rich text. Tags are found using a regex tag_matcher, and their
+    specific handler (eg PageLinkHandler) is looked up using a dict, handlers, which maps the value of type_attribute (where type_attribute might be emebedtype)
+    to a specific handler.
+    """
+    def __init__(self, handlers, tag_matcher, type_attribute, destination_ids_by_source={}):
+        self.handlers = handlers
+        self.tag_matcher = tag_matcher
+        self.type_attribute = type_attribute
+        self.destination_ids_by_source = destination_ids_by_source
+
+    def update_tag_id(self, match):
+        # Updates a specific tag's id from source to destination Wagtail instance
+        attrs = extract_attrs(match.group(1))
+        try:
+            handler = self.handlers[attrs[self.type_attribute]]
+            target_model = get_base_model(handler.get_model())
+            new_id = self.destination_ids_by_source.get((target_model, int(attrs['id'])), int(attrs['id']))
+            return FIND_ID.sub('id="{0}"'.format(str(new_id)), match.group(0))
+        except KeyError:
+            # If the relevant handler cannot be found, don't update the tag id
+            pass
+        return match.group(0)
+
+    def get_objects(self, html):
+        # Gets object references
+        objects = set()
+        for match in self.tag_matcher.finditer(html):
+            attrs = extract_attrs(match.group(1))
+            try:
+                handler = self.handlers[attrs[self.type_attribute]]
+                objects.add((get_base_model(handler.get_model()), int(attrs['id'])))
+            except KeyError:
+                # If no handler can be found, no object reference can be added.
+                # This might occur when the link is a plain url
+                pass
+        return objects
+
+    def update_ids(self, html):
+        # Update source instance ids to destination instance ids when possible
+        return self.tag_matcher.sub(self.update_tag_id, html)
+
+
+class MultiTypeRichTextReferenceHandler:
+    """Handles retrieving object references and updating ids for several different kinds of tags in rich text"""
+    def __init__(self, handlers):
+        self.handlers = handlers
+
+    def set_context_ids(self, destination_ids_by_source):
+        for handler in self.handlers:
+            handler.destination_ids_by_source = destination_ids_by_source
+
+    def update_ids(self, html):
+        for handler in self.handlers:
+            html = handler.update_ids(html)
+        return html
+
+    def get_objects(self, html):
+        objects = set()
+        for handler in self.handlers:
+            objects = objects.union(handler.get_objects(html))
+        return objects
+
+
+embed_handlers = features.get_embed_types()
```

Comment:
I'm not sure that it's safe to call `get_embed_types` / `get_link_types` at the top level of a module, as these functions trigger loading wagtail_hooks modules (and, given that this module will potentially get loaded while Django is iterating over INSTALLED_APPS, it might end up trying to load the wagtail_hooks module for an app that hasn't been loaded yet). `wagtail.core.rich_text` jumps through some hoops to [delay this until the first call to `expand_db_html`](https://github.com/wagtail/wagtail/blob/d65532dd0f0bf297523a4925fd79286b1d403a5d/wagtail/core/rich_text/__init__.py#L24) - I think we need to do something similar here.

Reply:
That's a good point. Thanks - I've changed over to a get_reference_handler method now which should be safer!